core/src/main/java/org/infinispan/container/DefaultDataContainer.java

```diff
@@ -318,48 +179,25 @@ public void clear() {
    }
 
    @Override
-   public Collection<V> values() {
-      return new Values();
-   }
-
-   @Override
-   public Set<InternalCacheEntry<K, V>> entrySet() {
-      return new EntrySet();
-   }
-
-   @Override
-   public void evict(K key) {
-      entries.computeIfPresent(key, (o, entry) -> {
-         passivator.passivate(entry);
-         return null;
-      });
+   public Iterator<InternalCacheEntry<K, V>> iterator() {
+      return new EntryIterator(entries.values().iterator());
    }
 
    @Override
-   public InternalCacheEntry<K, V> compute(K key, ComputeAction<K, V> action) {
-      return entries.compute(key, (k, oldEntry) -> {
-         InternalCacheEntry<K, V> newEntry = action.compute(k, oldEntry, entryFactory);
-         if (newEntry == oldEntry) {
-            return oldEntry;
-         } else if (newEntry == null) {
-            activator.onRemove(k, false);
-            return null;
-         }
-         activator.onUpdate(k, oldEntry == null);
-         if (trace)
-            log.tracef("Store %s in container", newEntry);
-         return newEntry;
-      });
+   public Iterator<InternalCacheEntry<K, V>> iterator(IntSet segments) {
+      return new FilterIterator<>(iterator(),
+            ice -> segments.contains(keyPartitioner.getSegment(ice.getKey())));
    }
 
    @Override
-   public Iterator<InternalCacheEntry<K, V>> iterator() {
-      return new EntryIterator(entries.values().iterator());
+   public Spliterator<InternalCacheEntry<K, V>> spliterator() {
+      return new EntrySpliterator(spliteratorIncludingExpired());
    }
 
    @Override
-   public Spliterator<InternalCacheEntry<K, V>> spliterator() {
-      return new EntrySpliterator(entries.values().spliterator());
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired() {
+      // Technically this spliterator is distinct, but it won't be set - we assume that is okay for now
```

Comment:
The only reason I made sure it was there, was because at some point I will be making the stream stuff care more about what flags are set and DISTINCT is just one more :D